File: Unity/Assets/Scripts/TileMap.cs
using System.Collections.Generic;
using UnityEngine;

namespace Gridia
{
    public class Tile
    {
        public int Floor { get; set; }
        public ItemInstance Item { get; set; }
        public Creature Creature { get; set; }
    }

    public class TileMap
    {
        private Tile[] tiles;
        private List<Creature> _creatures;
        public int Size { get; private set; }

        public TileMap (int size)
        {
            Size = size;//smell?
            _creatures = new List<Creature>();
            InitializeTiles ();
        }

        public Tile GetTile (int x, int y)
        {
            return tiles [Wrap (y) * Size + Wrap (x)];
        }

        public void SetFloor (int floor, int x, int y)
        {
            GetTile (x, y).Floor = floor;
        }

        public void SetItem (ItemInstance item, int x, int y)
        {
            GetTile (x, y).Item = item;
        }

        private Tile GetTileOfCreature(Creature creature) {
            return GetTile((int)creature.Position.x, (int)creature.Position.y);
        }

        public void AddCreature(Creature creature) {
            GetTileOfCreature(creature).Creature = creature;
            _creatures.Add(creature);
        }

        public void UpdateCreature(Creature creature, Vector2 newPosition) {
            GetTileOfCreature(creature).Creature = null;
            creature.Position = newPosition;
            GetTileOfCreature(creature).Creature = creature;
        }

        public bool Walkable(int x, int y) {
            return GetTile(x, y).Creature == null;
        }

        private void InitializeTiles ()
        {
            tiles = new Tile[Size * Size];
            for (int i = 0; i < tiles.Length; i++) {
                Tile tile = new Tile ();
                tile.Floor = 8;
                tile.Item = ContentManager.Singleton.GetItem(Random.Range(0, 10)).GetInstance();
                if (Random.Range(0, 10) > 1) tile.Item = ContentManager.Singleton.GetItem(0).GetInstance();
                //tile.Item = ContentManager.Singleton.GetItem(r.Next(ContentManager.Singleton.ItemCount)).GetInstance();
                tiles [i] = tile;
            }
        }

        private int Wrap (int value)
        {
            int mod = value % Size;
            return mod < 0 ? Size + mod : mod;
        }
    }
}

File: Server/GridiaServer/src/main/java/hoten/gridia/map/TileMap.java
package hoten.gridia.map;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import hoten.gridia.Creature;
import hoten.gridia.content.ItemInstance;
import hoten.serving.fileutils.FileUtils;
import java.io.File;

public class TileMap {

    public static TileMap loadMap(String mapName) {
        String json = FileUtils.readTextFile(new File(mapName + "/meta.json"));
        JsonObject metaData = new Gson().fromJson(json, JsonObject.class);

        int size = metaData.get("size").getAsInt();
        int depth = metaData.get("depth").getAsInt();
        int sectorSize = metaData.get("sectorSize").getAsInt();

        return new TileMap(mapName, size, depth, sectorSize, new JsonSectorLoader(), new SectorSaver());
    }

    public final int size, depth, sectorSize, area, volume, sectorsAcross, sectorsFloor, sectorsTotal;
    public final String mapName;
    private final Sector[][][] _sectors;
    private final SectorLoader _sectorLoader;
    private final SectorSaver _sectorSaver;

    public TileMap(String mapName, int size, int depth, int sectorSize, SectorLoader sectorLoader, SectorSaver sectorSaver) {
        if (size % sectorSize != 0) {
            throw new IllegalArgumentException("sectorSize must be a factor of size");
        }
        this.mapName = mapName;
        this.size = size;
        this.depth = depth;
        this.sectorSize = sectorSize;
        area = size * size;
        volume = area * depth;
        sectorsAcross = size / sectorSize;
        sectorsFloor = sectorsAcross * sectorsAcross;
        sectorsTotal = sectorsFloor * depth;
        _sectors = new Sector[sectorsAcross][sectorsAcross][depth];
        _sectorLoader = sectorLoader;
        _sectorSaver = sectorSaver;
    }

    //temporary
    public void loadAll() {
        for (int x = 0; x < sectorsAcross; x++) {
            for (int y = 0; y < sectorsAcross; y++) {
                for (int z = 0; z < depth; z++) {
                    Sector sector = _sectors[x][y][z];
                    if (sector != null) {
                        _sectors[x][y][z] = _sectorLoader.load(mapName, sectorSize, x, y, z);
                    }
                }
            }
        }
    }

    public void save() {
        for (int x = 0; x < sectorsAcross; x++) {
            for (int y = 0; y < sectorsAcross; y++) {
                for (int z = 0; z < depth; z++) {
                    Sector sector = _sectors[x][y][z];
                    if (sector != null) {
                        _sectorSaver.save(mapName, sector);
                    }
                }
            }
        }
    }

    public String getSectorData(int sx, int sy, int sz) {
        Sector sector = getSector(sx, sy, sz);
        Gson gson = new Gson(); // :(
        return gson.toJson(sector._tiles);
    }

    public Sector getSector(int sx, int sy, int sz) {
        Sector sector = _sectors[sx][sy][sz];
        if (sector == null) {
            _sectors[sx][sy][sz] = sector = _sectorLoader.load(mapName, sectorSize, sx, sy, sz);
        }
        return sector;
    }

    public Sector getSectorOf(Coord loc) {
        return getSectorOf(loc.x, loc.y, loc.z);
    }

    public Sector getSectorOf(int x, int y, int z) {
        x = wrap(x);
        y = wrap(y);
        int sx = x / sectorSize;
        int sy = y / sectorSize;
        return getSector(sx, sy, z);
    }

    public Tile getTile(Coord loc) {
        return getTile(loc.x, loc.y, loc.z);
    }

    public Tile getTile(int x, int y, int z) {
        x = wrap(x);
        y = wrap(y);
        return getSectorOf(x, y, z).getTile(x % sectorSize, y % sectorSize);
    }

    public int getFloor(int x, int y, int z) {
        return getTile(x, y, z).floor;
    }

    public void setFloor(int x, int y, int z, int floor) {
        getTile(x, y, z).floor = floor;
    }

    public void setFloor(Coord loc, int floor) {
        getTile(loc).floor = floor;
    }

    public ItemInstance getItem(Coord c) {
        return getTile(c.x, c.y, c.z).item;
    }

    public ItemInstance getItem(int x, int y, int z) {
        return getTile(x, y, z).item;
    }

    public void setItem(ItemInstance item, int x, int y, int z) {
        getTile(x, y, z).item = item;
    }

    public void setItem(ItemInstance item, Coord c) {
        getTile(c.x, c.y, c.z).item = item;
    }

    public Creature getCreature(Coord loc) {
        return getTile(loc).cre;
    }

    public int wrap(int value) {
        int mod = value % size;
        return mod < 0 ? size + mod : mod;
    }

    public Coord wrap(Coord loc) {
        return new Coord(wrap(loc.x), wrap(loc.y), loc.z);
    }

    public boolean walkable(int x, int y, int z) {
        Tile tile = getTile(x, y, z);
        return tile.cre == null && tile.item.data.walkable && tile.floor != 0;
    }

    public Coord getCoordFromIndex(int index) {
        int z = index / area;
        int y = (index / size) % size;
        int x = index % size;
        return new Coord(x, y, z);
    }

    public boolean inBounds(int x, int y, int z) {
        return x >= 0 && x < size && y >= 0 && y < size && z >= 0 && z < depth;
    }
}
